### PR TITLE
Use revocable memory in spillable HashAggregationOperator

### PR DESCRIPTION
**Pull request by pnowojski**
**As part of this PR please review only THREE last commits:**
1. Add setRevocableMemoryReservation method
2. Use revocable memory for spilling in aggregation
3. Add query.max-revocable-memory-per-node property

This pull request requires both https://github.com/prestodb/presto/pull/5826 and https://github.com/prestodb/presto/pull/5711.

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java [hunk 1]
```diff
@@ -286,6 +286,14 @@ public void addInput(Page page)
     }
 
     @Override
+    public void revokeMemory()
+    {
+        if (aggregationBuilder != null && !aggregationBuilder.isBusy()) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java]: move the `!aggregationBuilder.isBusy()` check to inside `aBuilder.revokeMemory()`. It is not obvious in this context that `isBusy` means `isSpillingAlready`.

File: presto-main/src/main/java/com/facebook/presto/execution/SqlTaskManager.java [hunk 2]
```diff
@@ -124,14 +124,21 @@ public SqlTaskManager(
...
                     throws Exception
             {
-                return new QueryContext(key, maxQueryMemoryPerNode, localMemoryManager.getPool(LocalMemoryManager.GENERAL_POOL), localMemoryManager.getPool(LocalMemoryManager.SYSTEM_POOL), taskNotificationExecutor);
+                return new QueryContext(
+                        key,
+                        maxQueryMemoryPerNode,
+                        localMemoryManager.getPool(LocalMemoryManager.GENERAL_POOL),
+                        localMemoryManager.getPool(LocalMemoryManager.SYSTEM_POOL),
+                        taskNotificationExecutor,
+                        maxQueryRevocableMemoryPerNode);
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlTaskManager.java]: Make it 3d argument

File: presto-main/src/main/java/com/facebook/presto/memory/NodeMemoryConfig.java [hunk 3]
```diff
@@ -24,8 +24,10 @@
 public class NodeMemoryConfig
 {
     public static final String QUERY_MAX_MEMORY_PER_NODE_CONFIG = "query.max-memory-per-node";
+    public static final String QUERY_MAX_REVOCABLE_MEMORY_PER_NODE_CONFIG = "query.max-revocable-memory-per-node";
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/memory/NodeMemoryConfig.java]: Make value of this property to follow value of `query.max-memory-per-node` as long only `query.max-memory-per-node` is specified in config.

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 4]
```diff
@@ -83,27 +80,30 @@ public void processPage(Page page)
...
     }
 
     public boolean checkFullAndUpdateMemory()
     {
-        systemMemoryContext.setBytes(hashAggregationBuilder.getSizeInMemory());
+        updateMemoryUsage();
         return false;
     }
 
+    private void updateMemoryUsage()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: I would inline this method

File: presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java [hunk 5]
```diff
@@ -83,27 +80,30 @@ public void processPage(Page page)
...
+    public void revokeMemory()
     {
-        return !spillInProgress.isDone();
+        CompletableFuture<?> completableFuture = spillToDisk();
+        completableFuture.thenRun(this::updateMemoryUsage);
     }
 
-    private boolean shouldSpill(long memorySize)
+    @Override
+    public boolean isBusy()
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/operator/aggregation/builder/SpillableHashAggregationBuilder.java]: maybe rename this as `isSplillnProgress`

**Comment by kokosing** [comment]
lgtm

File: presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java (hunk 1 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java]: done

File: presto-main/src/main/java/com/facebook/presto/execution/SqlTaskManager.java (hunk 2 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/execution/SqlTaskManager.java]: it has default value in another constructor, so I will keep it as it is

File: presto-main/src/main/java/com/facebook/presto/memory/NodeMemoryConfig.java (hunk 3 above)
Comment (pnowojski) [inline presto-main/src/main/java/com/facebook/presto/memory/NodeMemoryConfig.java]: done

**Pull request closed without merging**